### PR TITLE
k8s: Fetch-free CEP update logic

### PR DESCRIPTION
**Pull request by raybejjani**
CEP update controllers would always fetch, update then upload CEP
objects. In larger deployments this would cause undue load on
kube-apiserver because the same controller is the only writer of the
CEP, and can re-use the previous update's instance.
The new controller reuses the CEP object and handles unexpected errors
by recreating the CEP from scratch. This means there is a 1 period delay
between refetching and updating a CEP when that occurs.

One side-effect is that unless the local state changes, deleted CEPs are not restored. They can only be recovered when they change or the agent restarts.

Fixes https://github.com/cilium/cilium/issues/7237

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7242)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-missing-k8s

**Comment by raybejjani** [comment]
test-missed-k8s

File: pkg/k8s/endpointsynchronizer/cep.go
Comment (tgraf) [inline pkg/k8s/endpointsynchronizer/cep.go]: Is this correct? When `Update()` failed because the revision is outdated, it provides the latest revision ready for retry. Please add comments to describe when exactly Update() fails and how we deal with outdated revisions.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Review by aanm — COMMENTED** [review]
@tgraf This will decrease the number of GETs being done for the endpoints. I'll wait until this PR is merged before moving with the patch change.

File: pkg/k8s/endpointsynchronizer/cep.go
Comment (raybejjani) [reply pkg/k8s/endpointsynchronizer/cep.go]: You are right, it wasn't correct. I shuffled the code around to handle the revision error case specifically, and to be more clear on the order of operations.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: pkg/k8s/endpointsynchronizer/cep.go
Comment (tgraf) [inline pkg/k8s/endpointsynchronizer/cep.go]: I still struggle with this info message. Is it an expected condition, if so, debug. Is it unexpected and rare -> warning.

**Review by tgraf — CHANGES_REQUESTED** [review]
LGTM except for the info message

File: pkg/k8s/endpointsynchronizer/cep.go
Comment (raybejjani) [reply pkg/k8s/endpointsynchronizer/cep.go]: It is very unexpected. It would mean something else is updating the CEP and that should never happen (even the GC would only happen if the endpoint is deleted so this controller would be stopped anyway). I'll change it to a warning.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-missing-k8s

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
failures are not related with this change

**Pull request merged**
